GLMakie/assets/shader/lines.geom

```diff
@@ -9,804 +9,441 @@ struct Nothing{ //Nothing type, to encode if some variable doesn't contain any d
+        extrusion[1][0] = extrusion[1][1];
+    } else {
+        extrusion[1][1] = dot(miter_n2, v1.xy) / miter_offset2;
+        extrusion[1][0] = -extrusion[1][1];
     }
 
-    // scaling of uv.y due to different linewidths
-    // the padding for AA_THICKNESS should always have AA_THICKNESS width in uv
-    thickness_aa1 = g_thickness[1] / edge_scale + AA_THICKNESS;
-    thickness_aa2 = g_thickness[2] / edge_scale + AA_THICKNESS;
+    // TODO: maybe only for double sharp joints
+    // truncated could extrude
+    // start/end could extrude
+
+    // factor to reduce size of quad to make sure vertices do not pass each other
+    // i.e. avoid:
+    //  _______
+    //  '.   .'
+    //     x
+    //   '---'
+    // by stopping the line at x
+    vec2 shape_factor = vec2(
+        max(0.0, segment_length / max(segment_length, (halfwidth + AA_THICKNESS) * (extrusion[0][0] - extrusion[1][0]))), // -n
+        max(0.0, segment_length / max(segment_length, (halfwidth + AA_THICKNESS) * (extrusion[0][1] - extrusion[1][1])))  // +n
+    );
 
-    // Generate line segment
-    u1 *= px2uv;
-    u2 *= px2uv;
+    // Now we can scale them to the linewidth
+    extrusion = halfwidth * extrusion;
 
-    // Normal Version
-    generate_line_segment(
-        p1, miter_a, u1, thickness_aa1,
-        p2, miter_b, u2, thickness_aa2,
-        v1.xy, segment_length
-    );
+    // Generate static/flat outputs
 
-    // Debug - show each triangle
-    // generate_line_segment_debug(
-    //     p1, miter_a, u1, thickness_aa1,
-    //     p2, miter_b, u2, thickness_aa2,
-    //     v1.xy, segment_length
-    // );
+    // Set up pattern overwrites at joints if patterns are used. This "freezes"
+    // the pattern in either the on state (draw) or off state (no draw) to avoid
+    // fragmenting it around a corner.
+    vec2 adjustment = process_pattern(pattern, isvalid, extrusion, halfwidth);
 
-    return;
-}
+    // If adjustment != 0.0 we replace a joint by an extruded line, so we no longer
+    // need to shrink the line for the joint to fit.
+    if (adjustment[0] != 0.0 || adjustment[1] != 0.0)
+        shape_factor = vec2(1.0);
 
+    // limit range of distance sampled in prev/next segment
+    // this makes overlapping segments draw over each other when reaching the limit
+    f_discard_limit = vec2(
+        is_truncated[0] ? 0.0 : 1e12,
+        is_truncated[1] ? 0.0 : 1e12
+    );
 
+    // used to elongate sdf to include joints
+    // if start/end       don't elongate TODO: this may need ~AA_RADIUS elongation to cleanly close lines?
+    // if joint skipped   elongate to new length
+    // if normal joint    elongate a lot to let discard/truncation handle joint
+    f_extrusion = vec2(
+        !isvalid[0] ? min(AA_RADIUS, halfwidth) : (adjustment[0] == 0.0 ? 1e12 : abs(extrusion[0][0])),
+        !isvalid[3] ? min(AA_RADIUS, halfwidth) : (adjustment[1] == 0.0 ? 1e12 : abs(extrusion[1][0]))
+    );
```

Comment:
The `AA_RADIUS` here is supposed to make looping lines connected more smoothly along a straight section. The `min(AA_RADIUS, halfwidth)` is supposed to prevent them from extruding too far when connected in a corner. This might be something to look out for in tests and adjust some more.
This is also what makes lines slightly longer now.